File: packages/@intlayer/babel/src/dictionaryImports.ts
import type * as BabelTypes from '@babel/types';
import { getPathHash } from '@intlayer/engine/utils';

/**
 * How a dictionary reaches a call site.
 *
 * - `static`  — the compiled JSON is imported directly.
 * - `dynamic` — a generated per-locale loader is imported and resolved at
 *               runtime (Suspense).
 * - `fetch`   — same as `dynamic`, but the loader fetches remote content.
 */
export type ImportMode = 'static' | 'dynamic' | 'fetch';

/** Identifier suffix that also tells the injection step which directory to read. */
const IDENT_SUFFIX: Record<ImportMode, string> = {
  static: '',
  dynamic: '_dyn',
  fetch: '_fetch',
};

/**
 * Builds the generated identifier for a dictionary key.
 *
 * Replicates the xxHash64 → Base-62 algorithm used by the SWC version and
 * prefixes an underscore so the generated identifiers never collide with
 * user-defined ones.
 */
export const makeDictionaryIdent = (
  babelTypes: typeof BabelTypes,
  key: string,
  mode: ImportMode
): BabelTypes.Identifier =>
  babelTypes.identifier(`_${getPathHash(key)}${IDENT_SUFFIX[mode]}`);

/**
 * The dictionary imports one file needs, accumulated while its call sites are
 * rewritten and injected in one pass at the end.
 *
 * Both the native and the compat rewrites write here, so a dictionary read
 * through `useIntlayer('about')` and through `useTranslation('about')` in the
 * same file shares a single import.
 *
 * `dynamic` and `fetch` share one map because they occupy the same import
 * slot, distinguished only by their identifier suffix — mirroring
 * `InjectedImports` in the SWC plugin.
 */
export type DictionaryImportRegistry = {
  /** Dictionary key → identifier of the static JSON (or nested companion) import. */
  readonly staticImports: Map<string, BabelTypes.Identifier>;
  /** Dictionary key → identifier of the dynamic / fetch loader import. */
  readonly dynamicImports: Map<string, BabelTypes.Identifier>;
  /**
   * Returns the identifier for `key` in `mode`, creating and registering it on
   * first use.
   */
  identFor: (key: string, mode: ImportMode) => BabelTypes.Identifier;
};

/** Creates an empty registry for one source file. */
export const createDictionaryImportRegistry = (
  babelTypes: typeof BabelTypes
): DictionaryImportRegistry => {
  const staticImports = new Map<string, BabelTypes.Identifier>();
  const dynamicImports = new Map<string, BabelTypes.Identifier>();

  return {
    staticImports,
    dynamicImports,
    identFor: (key, mode) => {
      const map = mode === 'static' ? staticImports : dynamicImports;

      const existing = map.get(key);
      if (existing) return existing;

      const ident = makeDictionaryIdent(babelTypes, key, mode);
      map.set(key, ident);
      return ident;
    },
  };
};

File: packages/@intlayer/swc/src/dictionary_imports.rs
//! The dictionary imports one file needs, accumulated while its call sites are
//! rewritten and injected in one pass at the end.
//!
//! Both the native and the extra-caller rewrites allocate through this registry,
//! so a dictionary read through `useIntlayer("about")` and through
//! `useTranslation("about")` in the same file shares a single import.

use crate::ast::make_hashed_ident;
use std::collections::BTreeMap;
use swc_core::ecma::ast::Ident;

/// Per-call import mode. Dynamic and fetch resolve to the same helper but to
/// different generated loader directories.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ImportKind {
    Static,
    Dynamic,
    Fetch,
}

impl ImportKind {
    /// Parses the wire value of `importMode` / `dictionaryModeMap`.
    /// Unrecognised values are treated as `"static"` so a typo never silently
    /// promotes a dictionary to a dynamic loader.
    pub fn from_option(raw: Option<&str>) -> Option<Self> {
        match raw {
            Some("dynamic") => Some(ImportKind::Dynamic),
            Some("fetch") => Some(ImportKind::Fetch),
            Some(_) => Some(ImportKind::Static),
            None => None,
        }
    }

    /// Suffix appended to the generated import identifier, which also tells the
    /// import-injection step which directory the loader lives in.
    fn ident_suffix(self) -> &'static str {
        match self {
            ImportKind::Static => "",
            ImportKind::Dynamic => "_dyn",
            ImportKind::Fetch => "_fetch",
        }
    }

    /// Whether the call site receives a loader plus its dictionary key rather
    /// than a plain dictionary object.
    pub fn is_dynamic_helper(self) -> bool {
        !matches!(self, ImportKind::Static)
    }
}

/// Dictionary imports the transform decided to inject, in insertion order.
#[derive(Default)]
pub struct InjectedImports {
    /// Dictionary key → identifier of the static JSON (or nested companion) import.
    pub static_imports: BTreeMap<String, Ident>,
    /// Dictionary key → identifier of the dynamic / fetch loader import.
    pub dynamic_imports: BTreeMap<String, Ident>,
}

impl InjectedImports {
    /// Returns the identifier for `key` in the map matching `import_kind`,
    /// creating and registering it on first use. Dynamic and fetch identifiers
    /// share one map because they resolve to the same import slot,
    /// distinguished only by their `_dyn` / `_fetch` suffix.
    pub fn ident_for(&mut self, key: &str, import_kind: ImportKind) -> Ident {
        let map = match import_kind {
            ImportKind::Static => &mut self.static_imports,
            ImportKind::Dynamic | ImportKind::Fetch => &mut self.dynamic_imports,
        };

        if let Some(ident) = map.get(key) {
            return ident.clone();
        }

        let ident = make_hashed_ident(key, import_kind.ident_suffix());
        map.insert(key.to_string(), ident.clone());
        ident
    }
}
